File: packages/contracts/src/partyAttendance.ts
// Party attendance (NUBARCA-PARTY-CHECKIN-01).
//
// Who ARRIVED, recorded by the host. A third fact beside the RSVP (what a
// person declared) and the PartyParticipant (the anonymous browser at the
// party), bound to neither and never inferred from either: scanning the
// party's QR is not an arrival, and an arrival never changes an RSVP.
//
// The guest list is OPTIONAL. An open party has no groups and records everybody
// as an other arrival; a party with a guest list records the people on it by
// guest and anybody else as an other arrival. Nothing here names a party
// "type" — which kind a party is follows from which rows exist.
//
// Every route is the HOST's, under /api/parties/{partyId}. Neither the party's
// QR token nor a personal invitation token can reach any of it.

import { foldSearchText, type PartyRsvpStatus } from './partyRsvp.ts';

export const PARTY_ATTENDANCE_LIMITS = {
  /** Unicode code points, like every Party text limit. */
  name: 120,
} as const;

/**
 * Where an arrival was first recorded: by the host, or by the guest's own group
 * from its personal invitation ("Sono qui", while the party is live). Both write
 * the same fact, and the first one to succeed sets its moment and its source.
 */
export const PARTY_ATTENDANCE_SOURCES = ['owner', 'invitation'] as const;
export type PartyAttendanceSource = (typeof PARTY_ATTENDANCE_SOURCES)[number];

// ── Owner DTOs (party.access, owner-private) ────────────────────────────────

/**
 * ONE definition per count, separate from `PartyRsvpSummary`. `expectedPeople`
 * counts guests whose RSVP is attending (named or +1); `expectedArrived` those
 * of them who arrived and `expectedMissing` those who have not;
 * `unexpectedKnownGuests` guests on the list who arrived with a pending or
 * declined RSVP; `otherArrivals` people recorded who are not on the list;
 * `totalArrivals` every arrival of either kind.
 */
export interface PartyAttendanceSummary {
  expectedPeople: number;
  expectedArrived: number;
  expectedMissing: number;
  unexpectedKnownGuests: number;
  otherArrivals: number;
  totalArrivals: number;
}

export interface PartyAttendanceGuest {
  guestId: string;
  name: string;
  isAdditionalGuest: boolean;
  /** What the person declared. An arrival never changes it. */
  rsvpStatus: PartyRsvpStatus;
  /** Null: not recorded as arrived. */
  checkedInAt: string | null;
  /** Who recorded it first. Null exactly when `checkedInAt` is. */
  checkInSource: PartyAttendanceSource | null;
}

export interface PartyAttendanceGroup {
  groupId: string;
  label: string;
  /** Named guests in the host's order, then the group's +1s. */
  guests: PartyAttendanceGuest[];
}

/** Somebody recorded as arrived who is not on the guest list. */
export interface PartyAttendanceOtherGuest {
  id: string;
  name: string;
  checkedInAt: string;
  version: number;
}

export interface PartyAttendance {
  partyId: string;
  partyStatus: 'draft' | 'published' | 'live' | 'ended';
  /** Live and Ended: arrivals may be recorded and corrected. */
  canEdit: boolean;
  summary: PartyAttendanceSummary;
  groups: PartyAttendanceGroup[];
  /** Latest first. */
  otherGuests: PartyAttendanceOtherGuest[];
}

/** One "add". The id is minted once per add and reused for its retries. */
export interface PartyAttendanceOtherGuestCreate {
  name: string;
  clientRequestId: string;
}

export interface PartyAttendanceOtherGuestUpdate {
  name: string;
  version: number;
}

// ── Rules ───────────────────────────────────────────────────────────────────

/** Whether the party has a guest list at all — read from the rows, never stored. */
export function hasGuestList(attendance: Pick<PartyAttendance, 'groups'>): boolean {
  return attendance.groups.length > 0;
}

/**
 * "Altri arrivi": everybody who arrived without an attending RSVP — guests on
 * the list who had not confirmed, and people not on it. It is exactly what the
 * `unexpected` filter shows, so Arrivati = (Attesi − Mancano) + Altri arrivi.
 */
export function unexpectedArrivals(summary: PartyAttendanceSummary): number {
  return summary.unexpectedKnownGuests + summary.otherArrivals;
}

export const PARTY_ATTENDANCE_FILTERS = ['all', 'to_arrive', 'arrived', 'unexpected'] as const;
export type PartyAttendanceFilter = (typeof PARTY_ATTENDANCE_FILTERS)[number];

/**
 * The filters worth offering. With no guest list nobody is expected and
 * everybody listed has arrived, so there is nothing to filter by.
 */
export function attendanceFiltersFor(attendance: Pick<PartyAttendance, 'groups'>): PartyAttendanceFilter[] {
  return hasGuestList(attendance) ? [...PARTY_ATTENDANCE_FILTERS] : [];
}

export function guestMatchesAttendanceFilter(guest: PartyAttendanceGuest, filter: PartyAttendanceFilter): boolean {
  const arrived = guest.checkedInAt !== null;
  switch (filter) {
    case 'all': return true;
    case 'to_arrive': return guest.rsvpStatus === 'attending' && !arrived;
    case 'arrived': return arrived;
    case 'unexpected': return arrived && guest.rsvpStatus !== 'attending';
  }
}

/** An other arrival has arrived and was not expected, by definition. */
export function otherGuestMatchesAttendanceFilter(filter: PartyAttendanceFilter): boolean {
  return filter !== 'to_arrive';
}

/**
 * What the door sees for one search and one filter, over the list the page
 * already holds. The search reads a guest's name, the label of their group and
 * an other arrival's name — case- and accent-insensitive. A group appears when
 * at least one of its people does.
 */
export function visibleAttendance(
  attendance: Pick<PartyAttendance, 'groups' | 'otherGuests'>,
  query: string,
  filter: PartyAttendanceFilter,
): { groups: PartyAttendanceGroup[]; otherGuests: PartyAttendanceOtherGuest[] } {
  const needle = foldSearchText(query.trim());
  const matches = (value: string) => needle === '' || foldSearchText(value).includes(needle);
  const groups = attendance.groups
    .map((group) => ({
      ...group,
      guests: group.guests.filter((guest) =>
        guestMatchesAttendanceFilter(guest, filter) && (matches(guest.name) || matches(group.label))),
    }))
    .filter((group) => group.guests.length > 0);
  const otherGuests = otherGuestMatchesAttendanceFilter(filter)
    ? attendance.otherGuests.filter((other) => matches(other.name))
    : [];
  return { groups, otherGuests };
}

// ── Routes ──────────────────────────────────────────────────────────────────

export function partyAttendancePath(partyId: string): string {
  return `/api/parties/${partyId}/attendance`;
}
export function partyAttendanceGuestPath(partyId: string, guestId: string): string {
  return `${partyAttendancePath(partyId)}/guests/${guestId}`;
}
export function partyAttendanceOtherGuestsPath(partyId: string): string {
  return `${partyAttendancePath(partyId)}/other-guests`;
}
export function partyAttendanceOtherGuestPath(partyId: string, attendanceGuestId: string): string {
  return `${partyAttendanceOtherGuestsPath(partyId)}/${attendanceGuestId}`;
}

/** "Sono qui" for one of the group's OWN people, on its personal invitation token. */
export function partyInvitationAttendanceGuestPath(token: string, guestId: string): string {
  return `/api/party-invitations/${encodeURIComponent(token)}/attendance/guests/${guestId}`;
}

File: src/NubArca.Api/Domain/PartyAttendance.cs
namespace NubArca.Api.Domain;

// ATTENDANCE — who actually arrived, recorded by the host.
//
// Three facts about people meet at a party, and this file adds the third
// without touching the other two:
//
//   PartyRsvp        what a person DECLARED before the party
//   PartyParticipant the anonymous BROWSER that uses the public party
//   attendance       who the host saw ARRIVE
//
// They are allowed to disagree, and nothing reconciles them. A guest who
// declined and came anyway keeps "declined" as the historical answer and gains
// an arrival; a guest who confirmed and never came keeps "attending" and has no
// arrival. Scanning the party's QR is not an arrival: the phone that scanned it
// is a PartyParticipant, and nothing here binds one to the other — not by name,
// email, phone, QR, cookie, IP or device.
//
// The guest list is OPTIONAL. A party with no invitation group at all — an
// open party — records its arrivals as PartyAttendanceGuest rows only, and a
// party with a guest list records the people on it as PartyGuestAttendance and
// everybody else as PartyAttendanceGuest. There is no party "type": which of
// the two a party has is simply which rows exist.
//
// Every row is owner-private. It reaches no public party surface, no personal
// invitation, the TV, the game, print, face search, a link preview, a log line
// or an audit record (audit lines carry ids only).

/// <summary>
/// The arrival of a person who is on the guest list. One row = arrived; no row
/// = not recorded as arrived. The key IS the guest, so a person arrives once —
/// however many times, from however many places, and however concurrently.
///
/// <para>TWO SOURCES, ONE FACT. The host may record it from the workspace, and
/// the guest's own group may record it from its personal invitation while the
/// party is live ("Sono qui"). Both write this same row; whichever succeeds
/// first sets <see cref="CheckedInAt"/> and <see cref="Source"/>, and every
/// later check-in from either side changes neither.</para>
///
/// <para>Deliberately no status, no check-out, no participant id, no token,
/// no device and no IP: this records that somebody arrived, not where they are
/// now, and not which phone they carry.</para>
/// </summary>
public class PartyGuestAttendance
{
    /// <summary>Primary key and foreign key: one arrival per guest, never two.</summary>
    public Guid PartyGuestId { get; set; }

    /// <summary>When the arrival was first recorded. Kept by a repeated check-in.</summary>
    public DateTime CheckedInAt { get; set; }

    /// <summary>Who recorded it first — see <see cref="PartyAttendanceSources"/>.</summary>
    public string Source { get; set; } = PartyAttendanceSources.Owner;

    public DateTime CreatedAt { get; set; }
}

/// <summary>
/// Where an arrival was recorded from. Closed, and held by the database as well
/// as here: a future source — a door scanner, a kiosk — is a slice that widens
/// the constraint and decides what that source may do, not a free string.
/// </summary>
public static class PartyAttendanceSources
{
    /// <summary>The host, from the party's workspace.</summary>
    public const string Owner = "owner";

    /// <summary>The guest's own group, from its personal invitation, while the party is live.</summary>
    public const string Invitation = "invitation";

    public static bool IsKnown(string? source) => source is Owner or Invitation;
}

/// <summary>
/// A person the host recorded as arrived who is NOT a <see cref="PartyGuest"/>
/// — every arrival at an open party, and anybody at an invited party who was
/// not on its list. A name and a moment, and nothing else: no RSVP, no
/// invitation, no email, no capability and no participant.
///
/// <para><see cref="ClientRequestId"/> is minted by the host's browser once per
/// "add" and reused for that add's retries, so a double tap or a replayed
/// request names the person once. <see cref="Version"/> guards a rename.</para>
/// </summary>
public class PartyAttendanceGuest
{
    public Guid Id { get; set; }
    public Guid PartyId { get; set; }

    public string Name { get; set; } = string.Empty;

    /// <summary>The name folded by <see cref="PartySearchText"/>. Derived.</summary>
    public string SearchText { get; set; } = string.Empty;

    public Guid ClientRequestId { get; set; }

    public DateTime CheckedInAt { get; set; }

    public int Version { get; set; } = 1;

    public DateTime CreatedAt { get; set; }
    public DateTime UpdatedAt { get; set; }
}

/// <summary>When the host may record, correct and remove arrivals.</summary>
public static class PartyAttendancePolicy
{
    /// <summary>
    /// While the party is happening and after it: Ended stays writable so a
    /// forgotten or mistaken arrival can be corrected once the evening is over.
    /// Before the party nobody has arrived, and recording an arrival never
    /// publishes or starts a party.
    /// </summary>
    public static bool IsOpen(string? partyStatus) =>
        partyStatus is PartyStatuses.Live or PartyStatuses.Ended;
}

/// <summary>What a host may type, measured in Unicode code points like every Party text limit.</summary>
public static class PartyAttendanceLimits
{
    public const int MaxNameLength = 120;

    /// <summary>
    /// A ceiling against accidents and abuse, not a product limit: the whole
    /// list is one owner read, and no door records more people than this.
    /// </summary>
    public const int MaxOtherGuestsPerParty = 5000;
}
